### PR TITLE
Combine ntuple types with templated ExponentVector type

### PR DESCRIPTION
**Pull request by mahrud**
- changed clang-format style for references
- inlined all util.hpp converters
- adjusted headers of aring-* files
- combined ntuple types into ExponentVector<T>
- switched M2_arrayint to std::vector<int> in matrix-kbasis

Next up is a similar pull request for replacing `intarray` with `gc_vector` and combining `varpower` types. @mikestillman what do you think about ExponentMap, ExponentArray, ExponentStack, or ExponentPool for the name for the new templated type that would replace `varpower`?

### CODE REVIEW

**Comment by mahrud** [comment]
Just to give a heads up, this is the first part of 3 or 4 more pull requests.

**Review by mikestillman — COMMENTED** [review]
@mahrud @DanGrayson this one is ready to go, I just want to make sure that one type change was intentional (it is in the comments)

File: M2/Macaulay2/e/f4/ntuple-monomial.hpp [hunk 1]
```diff
@@ -3,260 +3,89 @@
...
-  return true;
-}
-
-inline unsigned long ntuple_monomials::mask(int nvars, const_ntuple_monomial a)
-{
-  unsigned long result = 0;
+  HashExponent result = 0;
   int i;
-  unsigned int j;
+  size_t j;
```
[the remark this answers is not in the document]
Comment (mahrud) [reply M2/Macaulay2/e/f4/ntuple-monomial.hpp:70]: I'm not sure if it was planned, but j is only used as a counter here so it should be fine. We should eventually discuss how to combine the `mask` methods, however.

[the remark this answers is not in the document]
Comment (mikestillman) [reply M2/Macaulay2/e/f4/ntuple-monomial.hpp:70]: I'll merge this now.  But this variable in f4/ntuple-monomial.hpp doesn't need to be a size_t, it is being used as a bit location in a word.

File: M2/Macaulay2/e/matrix-kbasis.cpp [hunk 2]
```diff
@@ -177,15 +184,12 @@ KBasis::KBasis(const Matrix *bottom,
...
     {
       kb_target_multidegree = D->make_one();
-      ntuple::copy(heft_vector->len, lo_degree, kb_target_multidegree);
-    }
-  else
-    {
-      kb_target_multidegree = 0;
+      ntuple::copy(mHeftVector.size(), lo_degree, kb_target_multidegree);
     }
+  else { kb_target_multidegree = 0; }
```
[the remark this answers is not in the document]
Comment (mikestillman) [reply M2/Macaulay2/e/matrix-kbasis.cpp:192]: No, don't worry.  We can address this later if needed.

**Pull request merged**